src/Compilers/Core/Portable/SourceGeneration/Nodes/GeneratorRunStateTable.cs

```diff
@@ -0,0 +1,120 @@
+            {
+                Debug.Assert(RecordingExecutedSteps);
+                Debug.Assert(table.HasTrackedSteps);
+                foreach (var step in table.Steps)
+                {
+                    RecordStepTree(step, addToOutputSteps: true);
+                }
+            }
+
+            [MemberNotNull(nameof(_namedSteps), nameof(_outputSteps))]
+            public bool RecordingExecutedSteps { get; }
+
+            public GeneratorRunStateTable ToImmutableAndFree()
+            {
+                if (!RecordingExecutedSteps)
+                {
+                    return new GeneratorRunStateTable(ImmutableDictionary<string, ImmutableArray<IncrementalGeneratorRunStep>>.Empty, ImmutableDictionary<string, ImmutableArray<IncrementalGeneratorRunStep>>.Empty);
+                }
+                return new GeneratorRunStateTable(StepCollectionToImmutableAndFree(_namedSteps), StepCollectionToImmutableAndFree(_outputSteps));
+            }
+
+            private static ImmutableDictionary<string, ImmutableArray<IncrementalGeneratorRunStep>> StepCollectionToImmutableAndFree(Dictionary<string, HashSet<IncrementalGeneratorRunStep>> builder)
+            {
+                ImmutableDictionary<string, ImmutableArray<IncrementalGeneratorRunStep>>.Builder resultBuilder = ImmutableDictionary.CreateBuilder<string, ImmutableArray<IncrementalGeneratorRunStep>>();
+
+                foreach (var stepsByName in builder)
+                {
+                    resultBuilder.Add(stepsByName.Key, stepsByName.Value.ToImmutableArrayOrEmpty());
+                }
+
+                return resultBuilder.ToImmutable();
+            }
+
+            private void RecordStepTree(IncrementalGeneratorRunStep step, bool addToOutputSteps)
+            {
+                Debug.Assert(RecordingExecutedSteps);
+                foreach (var (inputStep, _) in step.Inputs)
+                {
+                    RecordStepTree(inputStep, addToOutputSteps: false);
+                }
+
+                // If the step name is not null, we record it in the easily accessible collections.
+                // Otherwise, it will be available only through graph traversal.
+                if (step.Name is not null)
+                {
+                    AddToNamedStepCollection(_namedSteps, step);
+                    if (addToOutputSteps)
+                    {
+                        AddToNamedStepCollection(_outputSteps, step);
+                    }
+                }
+
+                static void AddToNamedStepCollection(Dictionary<string, HashSet<IncrementalGeneratorRunStep>> stepCollectionBuilder, IncrementalGeneratorRunStep step)
+                {
+                    Debug.Assert(step.Name is not null);
+                    if (stepCollectionBuilder.TryGetValue(step.Name, out var stepsByName))
+                    {
+                        if (!stepsByName.Contains(step))
+                        {
+                            stepsByName.Add(step);
```

Comment:
>stepsByName.Add(step); [](http://example.com/codeflow?start=28&length=22)

Consider moving this outside the `if (stepCollectionBuilder.TryGetValue(...)` block and execute unconditionally.